src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -0,0 +1,314 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import importlib.resources
+import json
+import logging
+from dataclasses import dataclass
+from typing import ClassVar, Iterable, Sequence, cast
+
+from pants.backend.python.target_types import UnrecognizedResolveNamesError
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.unions import UnionMembership, union
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    CoursierResolvedLockfile,
+)
+from pants.option.subsystem import Subsystem
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet
+
+DEFAULT_TOOL_LOCKFILE = "<default>"
+NO_TOOL_LOCKFILE = "<none>"
+
+
+logger = logging.getLogger(__name__)
+
+
+class JvmToolBase(Subsystem):
+    """Base class for subsystems that configure a set of artifact requirements for a JVM tool."""
+
+    # Default version of the tool. (Subclasses must set.)
+    default_version: ClassVar[str]
```

Comment:
The intent here is to provide a simpler way for users to configure the tool version. The version is then substituted into `default_artifacts` via a `%VERSION%` placeholder.

Reply:
There has been a growing convention around using Python's default templating for things: so just `{version}` see abb022bce703b38bf300c51389b723d8d4d6e127.